vignettes/target_types.Rmd

```diff
@@ -0,0 +1,49 @@
+  %\VignetteEngine{knitr::rmarkdown}
+  \usepackage[utf8]{inputenc}
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE)
+```
+
+`scipiper` targets can be indicator files with a [shared cache](shared_cache.html), local data files with or without local indicator files, or simply R objects.  This vignette provides some guidelines for when to use these different patterns.
+
+## Objectives
+There are two main goals to keep in mind when deciding what style of target to use:
+ 
+ - **Minimize time spend unnecessarily rebuilding targets** --- i.e., those that have already been built locally or by collaborators
+ - **Keeping the pipeline simple** --- caching can save lots of build time, but also adds to the code that has to be maintained, can clutter the repo with build and indicator files, etc.
+ 
+ These two can sometimes be in tension, but usually it will be clear which is more important for a given case.
+ 
+## Key factors to consider
+
+ - **Time to create an object** vs **object size**: This is the biggest thing to consider --- is it faster to upload a file to the shared cache and have others download it, or just have others rebuild it themselves?  This will somewhat depend on your internet connection.  Don't forget about upload time as well.  Uploading is generally slower than downloading, especially for remote workers with residential internet connections.   
+ 
+ - **How often will a target be updated?**:  Saving a target to disk may be less useful if it is likely to need to be rebuilt before the file on disk is ever used again.
+ 
+ - **Collaborators**: Are all collaborators equipped to rebuild every step in the pipeline, if needed?  If a particular step needs to be done by certain people, say a processing-intensive munging step early in the pipeline, than this file will almost certainly need to be in the remote cache.
+ 
+## Target choices
+
+**Shared cache data/indicator file**: The target will be only be rebuilt if its dependencies have changed, and the indicator file has not been updated to trigger a download from the remote cache. For longer-running operations that are worthwhile to cache remotely.  See the [shared caching vignette](shared_cache.html) for different implementation options.
+
+**Local data file only**:  The target will be rebuild if the local data file does not exist or depedencies have changed.  For cases where a target takes a relatively long time to build, but creates a massive file that is more cumbersome to upload/download than rebuild.  
+
+**Local data file with local indicator**: The target will be rebuilt if the indicator file is missing or dependencies have changed.  This is typically used to speed up `scipiper` by reducing redundant hashing of large files, or to represent multiple files, allowing a "one-to-many" target.
+
+**R object**: The target will be rebuilt every time.  For quick-building targets that aren't worth the time to write to disk, much less upload, e.g. most plots.  
```

Comment:
It's still only rebuilt every time it's missing or out of date on the local computer. We still get the standard remake benefits; it's just that every team member needs to build it locally once. This is another place where I think this document would be stronger if we could find a way to talk both about the question of shared cache vs unshared and the question of local file vs object.

Interesting that you single out plots. I think first of small data objects (e.g., configuration snippets, tiny data.frames, character vectors) that will be applied or combined to make something more broadly useful in subsequent steps. Or data that are changing really frequently as we write more code, going to your point about "How often will a target be updated?"

I actually think of plots as often being a really good thing to push to a shared cache, because often I want collaborators to be able to see and discuss the plots with me. Unless the collaborators are actively keeping up to speed with building a pipeline locally (definitely not the case for our non-R collaborators, and sometimes not the case even for R users), it's a lot quicker for us all to browse to a Drive folder than for everybody to build the plots locally.

Reply:
Mm good point — my point of view is mainly from building the analytics log processing pipelines, where sharing plots, and the plots need to be in memory so they can be passed to bookdown

Reply:
And I did not know that all objects get written to disk behind the scenes! I guess for local-only targets, it just comes down to whether you want easy access without going through remake.